src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.cs

```diff
@@ -0,0 +1,354 @@
+        ///  -------------------------------------------------------------------
+        ///  | DataId (primary key, integer) | | Checksum (blob) | Data (blob) |
+        ///  -------------------------------------------------------------------
+        /// </summary>
+        private const string DocumentDataTableName = "DocumentData" + Version;
+
+        private const string DataIdColumnName = "DataId";
+        private const string ChecksumColumnName = "Checksum";
+        private const string DataColumnName = "Data";
+
+        private readonly CancellationTokenSource _shutdownTokenSource = new CancellationTokenSource();
+
+        private readonly IDisposable _dbOwnershipLock;
+        private readonly IPersistentStorageFaultInjector _faultInjectorOpt;
+
+        // Accessors that allow us to retrieve/store data into specific DB tables.  The
+        // core Accessor type has logic that we to share across all reading/writing, while
+        // the derived types contain only enough logic to specify how to read/write from
+        // their respective tables.
+
+        private readonly SolutionAccessor _solutionAccessor;
+        private readonly ProjectAccessor _projectAccessor;
+        private readonly DocumentAccessor _documentAccessor;
+
+        // cached query strings
+
+        private readonly string _select_star_from_string_table;
+        private readonly string _insert_into_string_table_values_0;
+        private readonly string _select_star_from_string_table_where_0_limit_one;
+
+        // We pool connections to the DB so that we don't have to take the hit of 
+        // reconnecting.  The connections also cache the prepared statements used
+        // to get/set data from the db.  A connection is safe to use by one thread
+        // at a time, but is not safe for simultaneous use by multiple threads.
+        private readonly object _connectionGate = new object();
+        private readonly Stack<SqlConnection> _connectionsPool = new Stack<SqlConnection>();
+
+        public SQLitePersistentStorage(
+            string workingFolderPath,
+            string solutionFilePath,
+            string databaseFile,
+            IDisposable dbOwnershipLock,
+            IPersistentStorageFaultInjector faultInjectorOpt)
+            : base(workingFolderPath, solutionFilePath, databaseFile)
+        {
+            _dbOwnershipLock = dbOwnershipLock;
+            _faultInjectorOpt = faultInjectorOpt;
+
+            _solutionAccessor = new SolutionAccessor(this);
+            _projectAccessor = new ProjectAccessor(this);
+            _documentAccessor = new DocumentAccessor(this);
+
+            _select_star_from_string_table = $@"select * from {StringInfoTableName}";
+            _insert_into_string_table_values_0 = $@"insert into {StringInfoTableName}(""{DataColumnName}"") values (?)";
+            _select_star_from_string_table_where_0_limit_one = $@"select * from {StringInfoTableName} where (""{DataColumnName}"" = ?) limit 1";
```

Comment:
So these aren't also involved in the in-memory table?

Reply:
correct. the string table is dumped (and read) only from the on-disk db.  they cahnge rarely enough (and the bulk is computed right at the beginning) that there's no need for the write cache.

Reply:
(we also write them in bulk anyways).